crates/block-producer/src/batch_builder/batch.rs

```diff
@@ -74,14 +78,24 @@ impl TransactionBatch {
         // Consumed notes are also removed from the unauthenticated input notes set in order to avoid
         // consumption of notes with the same ID by one single input.
         //
+        // If a note isn't consumed in place, its nullifier is added to the produced nullifiers set.
+        //
         // One thing to note:
         // This still allows transaction `A` to consume an unauthenticated note `x` and output note `y`
         // and for transaction `B` to consume an unauthenticated note `y` and output note `x`
         // (i.e., have a circular dependency between transactions), but this is not a problem.
         let output_notes: Vec<_> = txs
             .iter()
             .flat_map(|tx| tx.output_notes().iter())
-            .filter(|&note| !unauthenticated_input_notes.remove(&note.id()))
+            .filter(|&note| {
+                let consumed_in_place = unauthenticated_input_notes.remove(&note.id());
+                if !consumed_in_place {
+                    if let Some(nullifier) = candidate_nullifiers.remove(&note.id()) {
+                        produced_nullifiers.push(nullifier);
+                    }
+                }
+                !consumed_in_place
+            })
             .cloned()
             .collect();
```

Comment:
I think this logic should works, but I wonder if a simpler way to do this would be to reverse the order. Specifically:

1. First, instead of iterating over input notes (as is done on lines 61 - 74 above), we iterate over output notes and build a map `BTreeMap<NoteId, OutputNote>`.
2. Then, we iterate over input notes to build a list of input notes for the batch. At the same time, we try to remove any unauthenticated note from the `BTreeMap`.
  a. Here, we can also perform the check that hashes of input/output notes match.
3. Lastly, the remaining values of the `BTeeMap` become the output notes of the batch.